recipes-core/initrdscripts/initramfs-stubdomain/init.sh

```diff
@@ -82,10 +82,122 @@ dmargs=$( xenstore-read $(xenstore-list -p "$vm_uuid/image/dm-argv" | sort ))
 echo "target $target vm_uuid $vm_uuid"
 echo "Invoking qemu with dmargs       = ${dmargs}"
 
+qemu_ifdown() {
+    nic_num=$1
+    echo "vif-$nic_num" > /sys/bus/xen/drivers/vif/unbind
+    xenstore-rm "device/vif/$nic_num"
+}
+
+mkdir /tmp/qmp
+mkfifo /tmp/qmp/qemu.in /tmp/qmp/qemu.out
+(
+  echo '{"execute": "qmp_capabilities"}' | tee /proc/self/fd/2
+) >/tmp/qmp/qemu.in &
+(
+set +x
+cat /tmp/qmp/qemu.out | tee /proc/self/fd/2 | \
+while IFS= read -r line; do
+    if [ $(echo "\\${line}" | grep -cim1 '"event": "DEVICE_DELETED", "data": {"device": "nic') -eq 1 ]; then
+        nic_num=$(echo "\\${line}" | sed 's/.*"event": "DEVICE_DELETED", "data": {"device": "nic\([[:digit:]]*\)",.*/\1/')
+        qemu_ifdown "$nic_num"
+    fi
+done
+) &
+
+qemu_ifup() {
+    # vifX.Y-emu  X=domid Y=devid
+    vif="$1"
+    num="${vif%-emu}"
+    num="${num#vif}"
+    devid="${num#*.}"
+    domid="${num%.*}"
+    br="br$devid"
+
+    my_domid=$( xenstore-read domid )
+    xs_mac=$( xenstore-read "/local/domain/$my_domid/device/vif/$devid/mac" )
+
+    for netif in $( ls /sys/class/net/ ); do
+        if [ "$xs_mac" == "$( cat "/sys/class/net/$netif/address" )" ]; then
+            eth="$netif"
+            break
+        fi
+    done
+
+    if [ -z "$eth" ] ; then
+        echo "qemu_ifup: Could not find interface with mac $xenstore_mac"
+        ls /sys/class/net
+
+        return 1
+    fi
+
+    if [ ! -e "/sys/class/net/$eth" ] ; then
+        echo "qemu_ifup: PV interface $eth does not exist."
+
+        return 1
+    fi
+
+    if [ -e "/sys/class/net/$br" ] ; then
+        echo "qemu_ifup: Bridge $br already exists"
```

Comment:
Should there be an early return here (or similar)? If the bridge exists, we still attempt to `addbr` down below. Looking at the original script, there are a few operations we opt-out of when the bridge is already present.

Reply:
I'll add a `return 1` for that case.  It's one of those "should never happen" things.